### PR TITLE
add operator skeleton

### PR DESCRIPTION
**Pull request by carlory**
**What type of PR is this?**

/kind feature

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/cc @lonelyCZ @RainbowMango

### CODE REVIEW

File: operator/pkg/scheme/scheme.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	operatorv1alpha1 "github.com/karmada-io/karmada/operator/pkg/apis/operator/v1alpha1"
+)
+
+// Scheme holds the aggregated Kubernetes's schemes and extended schemes.
+var Scheme = runtime.NewScheme()
```
Comment (carlory) [inline operator/pkg/scheme/scheme.go:12]: @lonelyCZ should I reuse the `gclient`package ?

https://github.com/karmada-io/karmada/blob/master/pkg/util/gclient/gclient.go#L40

Comment (lonelyCZ) [reply operator/pkg/scheme/scheme.go:12]: I think we only need to use the specific resource, what resources will it operate?

Comment (carlory) [reply operator/pkg/scheme/scheme.go:12]: The operator only requires the karmada and some kubernetes resource. I'm not sure if we reuse the existing scheme in gclient package.

Comment (lonelyCZ) [reply operator/pkg/scheme/scheme.go:12]: Avoiding confusion between karmada built-in resources and operators, I think we can redefine a client that only has necessary scheme.

and cc @RainbowMango

File: operator/pkg/controller/karmada/karmada_controller.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	// examine DeletionTimestamp to determine if object is under deletion
+	if karmada.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(karmada, ControllerFinalizerName) {
+			controllerutil.AddFinalizer(karmada, ControllerFinalizerName)
+			if err := ctrl.Update(ctx, karmada); err != nil {
+				return controllerruntime.Result{}, err
```
[the remark this answers is not in the document]
Comment (carlory) [reply operator/pkg/controller/karmada/karmada_controller.go:64]: @jwcesign The Controller will requeue the Request to be processed again if an error is non-nil. 

FYI: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.13.0/pkg/reconcile/reconcile.go#L102

File: operator/pkg/controller/karmada/karmada_controller.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+	klog.V(2).InfoS("Reconciling karmada", "name", req.Name)
+
+	// do reconcile
+
+	return controllerruntime.Result{}, nil
+}
+
+func (ctrl *Controller) deleteUnableGCResources(karmada *operatorv1alpha1.Karmada) error {
+	klog.InfoS("Deleting unable gc resources", "karmada", klog.KObj(karmada))
```
[the remark this answers is not in the document]
Comment (carlory) [reply operator/pkg/controller/karmada/karmada_controller.go:95]: When we use statefulsets to implement a highly available etcd cluster, we need to clean up its associated pvc resources in this function.

File: operator/pkg/controller/karmada/karmada_controller.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		if errors.IsNotFound(err) {
+			klog.V(2).InfoS("Karmada has been deleted", "karmada", req)
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{}, err
+	}
+
+	// Deep-copy otherwise we are mutating our cache.
+	// TODO: Deep-copy only when needed.
+	karmada = karmada.DeepCopy()
```
Comment (RainbowMango) [inline operator/pkg/controller/karmada/karmada_controller.go:54]: I don't think the variable `karmada` now using cache. It points to an initialized memory by `&operatorv1alpha1.Karmada{}`.

File: operator/pkg/controller/karmada/karmada_controller.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+			controllerutil.AddFinalizer(karmada, ControllerFinalizerName)
+			if err := ctrl.Update(ctx, karmada); err != nil {
+				return controllerruntime.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(karmada, ControllerFinalizerName) {
+			// our finalizer is present, so lets handle any external dependency
+			if err := ctrl.deleteUnableGCResources(karmada); err != nil {
```
Comment (RainbowMango) [inline operator/pkg/controller/karmada/karmada_controller.go:71]: Can you remind me which resources can't be GCed here?

File: operator/pkg/scheme/scheme.go (hunk 1 above)
Comment (RainbowMango) [reply operator/pkg/scheme/scheme.go:12]: makes sense to me.

File: operator/pkg/controller/karmada/karmada_controller.go (hunk 5 above)
Comment (carlory) [reply operator/pkg/controller/karmada/karmada_controller.go:71]: Please see https://github.com/karmada-io/karmada/pull/2973#discussion_r1052887651

File: operator/pkg/controller/karmada/karmada_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply operator/pkg/controller/karmada/karmada_controller.go:91]: Got it.

**Comment by RainbowMango** [comment]
LGTM. @lonelyCZ do you have any other comments?

**Review by RainbowMango — APPROVED** [review]
/approve

**Pull request merged**